src/main/java/org/dataloader/orchestration/Orchestrator.java

```diff
@@ -0,0 +1,104 @@
+     *
+     * @param dataLoader the data loader to start with
+     * @param <K>        the key type
+     * @param <V>        the value type
+     * @return a new {@link Orchestrator}
+     */
+    public static <K, V> Orchestrator<K, V> orchestrate(DataLoader<K, V> dataLoader) {
+        return new Orchestrator<>(new Tracker(), dataLoader);
+    }
+
+    public Tracker getTracker() {
+        return tracker;
+    }
+
+    private Orchestrator(Tracker tracker, DataLoader<K, V> dataLoader) {
+        this.tracker = tracker;
+        this.startingDL = dataLoader;
+    }
+
+
+    public Step<K, V> load(K key) {
+        return load(key, null);
+    }
+
+    public Step<K, V> load(K key, Object keyContext) {
+        return Step.loadImpl(this, castAs(startingDL), key, keyContext);
+    }
+
+    static <T> T castAs(Object o) {
+        //noinspection unchecked
+        return (T) o;
+    }
+
+
+    <KT, VT> void record(Step<KT, VT> step) {
+        steps.add(step);
+        tracker.incrementStepCount();
+    }
+
+    /**
+     * This is the callback point saying to start the DataLoader loading process.
+     * <p>
+     * The type of object returned here depends on the value type of the last Step.  We cant be truly generic
+     * here and must be case.
+     *
+     * @param <VT> the value type
+     * @return the final composed value
+     */
+    <VT> CompletableFuture<VT> execute() {
+        Assertions.assertState(!steps.isEmpty(), () -> "How can the steps to run be empty??");
+        int index = 0;
+        Step<?, ?> firstStep = steps.get(index);
+
+        CompletableFuture<Object> currentCF = castAs(firstStep.codeToRun().apply(null)); // first load uses variable capture
+        whenComplete(index, firstStep, currentCF);
+
+        for (index++; index < steps.size(); index++) {
+            Step<?, ?> nextStep = steps.get(index);
+            Function<Object, CompletableFuture<?>> codeToRun = castAs(nextStep.codeToRun());
+            CompletableFuture<Object> nextCF = currentCF.thenCompose(value -> castAs(codeToRun.apply(value)));
+            currentCF = nextCF;
+
+            // side effect when this step is complete
+            whenComplete(index, nextStep, nextCF);
+        }
+        return castAs(currentCF);
+
+    }
+
+    private void whenComplete(int index, Step<?, ?> step, CompletableFuture<Object> cf) {
+        cf.whenComplete((v, throwable) -> {
+            getTracker().loadCallComplete(step.dataLoader());
+            // replace with instrumentation code
+            if (throwable != null) {
+                // TODO - should we be cancelling future steps here - no need for dispatch tracking if they will never run
+                System.out.println("A throwable has been thrown on step  " + index + ": " + throwable.getMessage());
+                throwable.printStackTrace(System.out);
+            } else {
+                System.out.println("step " + index + " returned : " + v);
+            }
+        });
+    }
```

Comment:
This is where we can have tracker side effects to know things.

We should probably also handle exceptions by cancelling all "expectations on future load" since they will not happen

Maybe we just mark the Tracker as "exceptional"